Lib/string.py

```diff
@@ -141,6 +141,30 @@ def convert(mo):
                              self.pattern)
         return self.pattern.sub(convert, self.template)
 
+    def is_valid(self):
+        for mo in self.pattern.finditer(self.template):
```

Comment:
Do you think you can refactor all the common code in these methods?  There's a lot of duplication, although I acknowledge it's imperfect.

Reply:
Unfortunately, I don't think so, at least not without rethinking how it works. All the methods do have a similar structure:
```
named = mo.group('named') or mo.group('braced')
if named is None:
  # do something particular to the method
elif mo.group('escaped') is not None:
  # use delimiter in substitute, do nothing in is_valid and get_identifier
elif mo.group('invalid') is not None:
  # do something particular to the method
else:
  raise ValueError
```
But they vary on which groups they care about and what to do with them.
I could do something where there's one common helper method that has the if-else chain, and each method passes in lambda expressions for the actions to take, but it seems clumsy and probably even harder to follow.

I've added some comments to the code to hopefully make it clearer.

Reply:
Awesome, thanks.  That's what I suspected.